### DISCUSSION TITLE
Help! I'm not sure  if I can use this library for my case (all selectors must be dynamic)

### DISCUSSION THREAD

**Discussion by Laura-Urb** [react-querybuilder/react-querybuilder#348] [Q&A] [answered]
Hi!
I have a complex user case in my app.
In my case,  intead of just a field selector with constants operators, I have a lists that change and  depend each other.
I have an extra selector that determinate wich obejct's array goin to be in my fields seletor. So, my first problem in that I think the library is not thinking to use a dynamic list on that selector, right?
My second problem is that the selected field going to tell me which list of combinators loads and the  value's unit.
Also I have to be able to load that extremmely dinamyc  query that iI submitd and be able to edited it
Is there any chanse that be able to do that?
Sorry for my poor english
Thanks
(I know I'm saving lists and load it as a totally wrong way)
**General Idea**
[option-1, option-2,....option-n] > selects one and loads dynamic list> [field-1,field-2,field-3,...field-n] > selects one and loads dynamic list >  [operator-1, operator-2...operator-n] 
![image](https://user-images.githubusercontent.com/63360295/180756179-d45783da-aee4-43d4-9b3a-b5933084cc91.png)
![image](https://user-images.githubusercontent.com/63360295/180756870-f322d71b-f1e2-4146-bc4d-629770392ae5.png)
![image](https://user-images.githubusercontent.com/63360295/180762301-90612699-fdd1-4dff-bc96-1cab59cbcb8d.png)

**Comment by jakeboone02** [comment, marked as answer]
Yes, you can meet those requirements with this library. #327 has a couple of example codesandboxes like [this one](https://codesandbox.io/s/uzo4lv), but they're probably a little over-engineered for what you need.

See if [this codesandbox](https://codesandbox.io/s/nice-torvalds-9jgp3b?file=/src/App.tsx) is something more like what you want. The relevant code is also below.

Since the `fields` prop can take an `OptionGroup<Field>[]` as well as `Field[]`, I set up a custom component to take advantage of that hierarchical structure. In the `CustomFieldSelector` I rendered one `<select>` to show the group label and another `<select>` to show the actual selected field (but with its options filtered by the appropriate group).

```tsx
import { useState } from 'react';
import { Field, FieldSelectorProps, OptionGroup, QueryBuilder, RuleGroupType } from 'react-querybuilder';
import 'react-querybuilder/dist/query-builder.scss';
import './styles.scss';

const fields: OptionGroup<Field>[] = [
  {
    label: 'Name',
    options: [
      { name: 'firstName', label: 'First Name' },
      { name: 'lastName', label: 'Last Name' },
    ],
  },
  {
    label: 'Contact',
    options: [
      { name: 'email', label: 'Email' },
      { name: 'phone', label: 'Phone' },
    ],
  },
];

const initialQuery: RuleGroupType = {
  combinator: 'and',
  rules: [
    { field: 'firstName', operator: '=', value: '' },
    { field: 'phone', operator: '=', value: '' },
  ],
};

const CustomFieldSelector = (props: FieldSelectorProps) => {
  const optGroups = props.options as OptionGroup<Field>[];
  const grp =
    optGroups.find((og) =>
      og.options.find((opt) => opt.name === props.value)
    ) ?? optGroups[0];

  return (
    <div className={props.className} style={{ display: 'flex', gap: '0.5rem' }}>
      <select
        value={grp.label}
        onChange={(e) =>
          props.handleOnChange(
            optGroups.find((og) => og.label === e.target.value)?.options[0].name
          )
        }
      >
        {optGroups.map((og) => (
          <option key={og.label} value={og.label}>
            {og.label}
          </option>
        ))}
      </select>
      <select
        value={props.value}
        onChange={(e) => props.handleOnChange(e.target.value)}
      >
        {grp.options.map((opt) => (
          <option key={opt.name} value={opt.name}>
            {opt.label}
          </option>
        ))}
      </select>
    </div>
  );
};

export default function () {
  const [query, setQuery] = useState(initialQuery);

  return (
    <div>
      <QueryBuilder
        fields={fields}
        query={query}
        onQueryChange={(q) => setQuery(q)}
        controlElements={{ fieldSelector: CustomFieldSelector }}
      />
    </div>
  );
}
```

Hope that helps!

**Reply by jakeboone02** [reply]
Hi @Laura-Urb, if the response above helped you out, could you mark it as the answer?
